src/server/tiered_storage.cc

```diff
@@ -184,10 +215,22 @@ void TieredStorage::ShardOpManager::Defragment(tiering::DiskSegment segment, str
       continue;
 
     stats_.total_defrags++;
-
-    // Cut out relevant part of value and restore it to memory
-    string_view value = page.substr(item_segment.offset - segment.offset, item_segment.length);
-    Upload(dbid, value, true, item_segment.length, &it->second);
+    PrimeValue& pv = it->second;
+    if (pv.IsCool()) {
+      PrimeValue::CoolItem item = pv.GetCool();
+      tiering::DiskSegment segment = FromCoolItem(item);
+
+      // We remove it from both cool storage and the offline storage.
+      PrimeValue hot = ts_->cool_queue_.Erase(item.record);
+      pv = std::move(hot);
+      auto* stats = GetDbTableStats(dbid);
+      stats->tiered_entries--;
+      stats->tiered_used_bytes -= segment.length;
+    } else {
+      // Cut out relevant part of value and restore it to memory
+      string_view value = page.substr(item_segment.offset - segment.offset, item_segment.length);
+      Upload(dbid, value, true, item_segment.length, &pv);
+    }
```

Comment:
I think this still could have been a part of Upload()

Reply:
what do you mean?